File: server/exec.go
/*
Copyright 2026 Orhan Yavuz.

This program is free software: you can redistribute it and/or modify
it under the terms of the GNU Affero General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU Affero General Public License for more details.

You should have received a copy of the GNU Affero General Public License
along with this program.  If not, see <https://www.gnu.org/licenses/>.
*/

package server

import (
	"context"
	"encoding/json"
	"errors"
	"fmt"
	"io"
	"log/slog"
	"net/http"
	"slices"
	"strings"
	"sync"
	"time"

	corev1 "k8s.io/api/core/v1"
	apierrors "k8s.io/apimachinery/pkg/api/errors"
	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
	"k8s.io/client-go/kubernetes"
	"k8s.io/client-go/kubernetes/scheme"
	"k8s.io/client-go/rest"
	"k8s.io/client-go/tools/remotecommand"
	utilexec "k8s.io/client-go/util/exec"

	"github.com/damgahq/damga/authz"
	"github.com/damgahq/damga/placement"
)

const (
	// maxCommandArgs and maxCommandArg bound what a request may ask to run.
	// Not a security control — anybody who reaches this endpoint may already
	// run a shell — but a bound on what this process will hold in memory
	// before it has decided anything.
	maxCommandArgs = 64
	maxCommandArg  = 4 << 10

	// execHeartbeat is why this streams rather than answering in one piece.
	//
	// The named case for this endpoint is "run a migration", which is the case
	// that takes minutes and prints nothing while it does. A buffered response
	// to a request that silent is closed by ingress-nginx after 60 seconds
	// (see logHeartbeat, which exists for the same reason) — and the command
	// has already run by then, so the caller loses the result of something
	// that did happen. That is the one outcome worth engineering against here.
	execHeartbeat = 20 * time.Second
)

// Errors this endpoint tells apart.
//
// One error for "exec failed" would be true of all of them at once, and each
// of these has a different next move: deploy the app, wait, say which
// container, fix the name, send a command, grant a permission. A message that
// cannot say which one is a message the reader has to guess from.
var (
	// errNoCommand: the request body carried no argv.
	errNoCommand = errors.New("no command was given: send {\"command\": [...]} with at least one element")

	// errNoRunnablePod: pods exist and none of them can take a command. The
	// message names the pods and what each is doing, because "not running" is
	// where a crash loop, an image that will not pull and a pod still being
	// created all look identical.
	errNoRunnablePod = errors.New("no pod is running")

	// errAmbiguousContainer: the pod has several containers and the request
	// named none. Refused rather than guessed — picking the first would run a
	// migration in a sidecar on the day somebody adds one.
	errAmbiguousContainer = errors.New("this pod has more than one container")

	// errNoSuchContainer: the named container is not in the pod.
	errNoSuchContainer = errors.New("no such container")
)

// ExecTarget is one command, in one container, of one pod.
type ExecTarget struct {
	// Namespace comes from the placement row, never from a convention, for the
	// reason placement carries it as a field.
	Namespace string
	App       string
	// Container may be empty, and the selection below decides what that means:
	// the only container, or a refusal naming the choices.
	Container string
	Command   []string
}

// Execer runs a command in a container and streams what it wrote.
//
// A seam for the reason LogSource is one: this is another place the control
// plane reaches into the cluster, and an install with no cluster to reach must
// still start, serve, and say so rather than appearing broken.
//
// Deliberately not a field on Options, and that is the difference between this
// and Builds. Builds is filled by the composition root because an installation
// may legitimately answer builds a different way; this is resolved from the
// pod's own ServiceAccount at first use and from nothing else. See
// inClusterExec. Nothing that constructs Options gains a field, so no existing
// call site quietly means something new by leaving one empty.
type Execer interface {
	// Exec returns the command's exit code. A non-zero code is a result and
	// not an error: a migration that exits 1 ran, and the caller needs to be
	// told what it printed and what it returned.
	Exec(ctx context.Context, t ExecTarget, stdout, stderr io.Writer) (int, error)
}

// execRoute runs one command in a tenant's container.
//
// # Why one command and not a terminal
//
// A browser terminal needs a terminal emulator, and the panel has no build
// step and no npm — panel.go states that as a decision and says it will not
// grow one. Shipping xterm.js would cost this repository the property that
// its front end has one dependency tree, in exchange for interactivity that
// the case this was asked for does not need: "run a migration" is one command
// whose output you read.
//
// # Why it streams
//
// See execHeartbeat. The command runs whether or not the connection survives,
// so the design question is whether the caller finds out what it did.
//
// # Why this is not an evidence record
//
// It was the first thing tried, and it is wrong for a reason worth writing
// down rather than rediscovering. An evidence Record is a deploy: its Seq
// means "the 41st deploy of api/prod", its Source is a commit, its Image is
// what that commit shipped, and its Hash chains to the record before it so
// that Verify can walk one Ref and say the history is intact. An exec has
// none of those. It carries no commit and no image, it has no lifecycle to
// transition through, and nothing about it can be replayed or verified later
// — the container it touched is gone by the next sync.
//
// Appending one anyway would do two kinds of damage. It would make Seq stop
// counting deploys, so "the 41st deploy" becomes a number that includes the
// times somebody opened a shell. And it would put a link in the hash chain
// that Verify has to accept but cannot check, which is worse than not
// recording it: a chain whose links mean two different things is a chain that
// proves less than it appears to.
//
// So it is a log line — see the two slog calls in execWith — and that is a
// weaker record, honestly weaker. It is not queryable per tenant, it is not
// tamper-evident, and it is not exported with the rest. What it is not is a
// deploy record that lies about what it counts. A tenant-visible audit trail
// for actions that are not deploys is a second store and a real piece of
// work; this endpoint is the second caller that wants one (the first is
// member:invite), and when there is a third it should be built rather than
// borrowed from the chain.
func execRoute(g guard, st stores) http.Handler {
	// Resolved once and reused, for the reason logs does the same: building a
	// client parses a token and a CA bundle off disk.
	return execWith(g, st, sync.OnceValues(inClusterExec))
}

// execWith is execRoute with the runner injected, which is how a test drives
// it without a cluster.
func execWith(g guard, st stores, open func() (Execer, error)) http.Handler {
	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		sub, ref, ok := g.admit(w, r, authz.ActionAppExec)
		if !ok {
			return
		}

		cmd, container, err := execRequestFrom(w, r)
		if err != nil {
			problem(w, http.StatusBadRequest, err.Error())
			return
		}

		place, err := st.placement.Get(r.Context(), ref.TenantID, ref.App, ref.Env)
		switch {
		case errors.Is(err, placement.ErrNotFound):
			problem(w, http.StatusNotFound, "this app and environment have no repository configured yet")
			return
		case err != nil:
			problem(w, http.StatusInternalServerError, "reading the placement failed")
			return
		}

		runner, err := open()
		if err != nil {
			// The shape createBuild uses, and for the same reason: a control
			// plane outside a cluster cannot exec into a pod, and saying so is
			// different from saying the command failed.
			slog.Warn("exec is unavailable", "error", err)
			problem(w, http.StatusNotImplemented,
				"this installation cannot run commands in the cluster")
			return
		}

		// Recorded before it runs, and again with what it returned. Which of
		// those two lines exists tells a reader whether a command that is not
		// accounted for was refused or was cut off mid-flight.
		//
		// argv[0] and the argument count, never the whole command line. A
		// command line is where people put a password — `psql -W hunter2`, a
		// token in an environment assignment — and this log has no access
		// control of its own, so recording all of it would move a tenant's
		// credentials somewhere their tenancy does not reach. What is kept is
		// enough to answer who ran something and roughly what: the identity,
		// the app, the container and the program.
		//
		// See the note on evidence in this file for why it is a log line and
		// not a record in the chain.
		slog.Info("exec started",
			"actor", sub.ID, "email", sub.Email, "tenant", ref.TenantID,
			"app", ref.App, "env", ref.Env, "namespace", place.Namespace,
			"container", container, "program", cmd[0], "args", len(cmd)-1)

		code, err := streamExec(w, r, runner, ExecTarget{
			Namespace: place.Namespace, App: ref.App,
			Container: container, Command: cmd,
		})
		slog.Info("exec finished",
			"actor", sub.ID, "tenant", ref.TenantID, "app", ref.App, "env", ref.Env,
			"program", cmd[0], "code", code, "error", err)
	})
}

// execRequestFrom reads the command and the optional container from the body.
func execRequestFrom(w http.ResponseWriter, r *http.Request) ([]string, string, error) {
	var body struct {
		Command   []string `json:"command"`
		Container string   `json:"container"`
	}
	dec := json.NewDecoder(http.MaxBytesReader(w, r.Body, maxRequestBody))
	dec.DisallowUnknownFields()
	if err := dec.Decode(&body); err != nil {
		return nil, "", fmt.Errorf("reading the request: %w", err)
	}
	if len(body.Command) == 0 {
		return nil, "", errNoCommand
	}
	if len(body.Command) > maxCommandArgs {
		return nil, "", fmt.Errorf("the command has %d arguments and the limit is %d",
			len(body.Command), maxCommandArgs)
	}
	for i, arg := range body.Command {
		if len(arg) > maxCommandArg {
			return nil, "", fmt.Errorf("argument %d is %d bytes and the limit is %d",
				i, len(arg), maxCommandArg)
		}
	}
	// An empty first element is a command that resolves to nothing, and the
	// kubelet's answer to it names neither the argument nor the request.
	if strings.TrimSpace(body.Command[0]) == "" {
		return nil, "", errNoCommand
	}
	return body.Command, body.Container, nil
}

// streamExec runs the command and writes what it produced as an event stream.
//
// Everything after the first byte is an event and never a status code, for the
// reason streamSSE says: once the headers are on the wire the status is spent,
// and a stream that reports a failure by stopping is one the reader cannot
// tell from a command that printed nothing.
func streamExec(w http.ResponseWriter, r *http.Request, runner Execer, t ExecTarget) (int, error) {
	h := w.Header()
	h.Set("Content-Type", "text/event-stream")
	h.Set("Cache-Control", "no-cache, no-transform")
	h.Set("Connection", "keep-alive")
	// The stream is a tenant's own output; no proxy on the way should hold a
	// copy of it.
	h.Set("X-Accel-Buffering", "no")
	w.WriteHeader(http.StatusOK)

	flusher, _ := w.(http.Flusher)
	out := &sseWriter{w: w, flusher: flusher}

	// The heartbeat and the command's own output share one writer, so the
	// ticker cannot interleave a beat into the middle of a chunk.
	ctx, cancel := context.WithCancel(r.Context())
	defer cancel()
	var beat sync.WaitGroup
	beat.Go(func() {
		tick := time.NewTicker(execHeartbeat)
		defer tick.Stop()
		for {
			select {
			case <-ctx.Done():
				return
			case <-tick.C:
				out.beat()
			}
		}
	})

	code, err := runner.Exec(ctx, t, out.stream("stdout"), out.stream("stderr"))
	cancel()
	beat.Wait()

	if err != nil {
		out.event("error", map[string]string{"message": execMessage(err, t)})
		return code, err
	}
	out.event("exit", map[string]any{"code": code})
	return code, nil
}

// execMessage turns a failure into a sentence that names which failure it was.
//
// The API server's own answer to a missing permission is "pods \"x\" is
// forbidden", which reads as the pod refusing rather than as this installation
// never having been granted the right. That distinction is the whole reason
// this function exists: one of these is fixed by an operator editing RBAC and
// the rest are not.
func execMessage(err error, t ExecTarget) string {
	switch {
	case errors.Is(err, ErrNoPods):
		return fmt.Sprintf("%s has no pods in %s: nothing is deployed, or it is scaled to zero",
			t.App, t.Namespace)
	case errors.Is(err, errNoRunnablePod), errors.Is(err, errAmbiguousContainer),
		errors.Is(err, errNoSuchContainer):
		// These already name themselves and what they saw.
		return err.Error()
	case apierrors.IsForbidden(err):
		return "this control plane may not run commands in pods. Its ClusterRole " +
			"grants pods and pods/log, and pods/exec has to be added to it — " +
			"see cluster/control-plane.yaml"
	case apierrors.IsNotFound(err):
		return "the pod was there when it was chosen and is not there now; it was " +
			"probably replaced mid-command"
	default:
		return err.Error()
	}
}

// sseWriter serializes everything written to one event stream.
//
// stdout and stderr are two io.Writers over one connection, and remotecommand
// writes to both from different goroutines. Without the lock a chunk of stderr
// lands inside a stdout event and the reader sees neither.
type sseWriter struct {
	mu      sync.Mutex
	w       http.ResponseWriter
	flusher http.Flusher
}

func (s *sseWriter) event(name string, payload any) {
	s.mu.Lock()
	defer s.mu.Unlock()
	_ = writeSSE(s.w, name, payload)
	if s.flusher != nil {
		s.flusher.Flush()
	}
}

// beat keeps the connection observably alive through a command that is working
// and saying nothing. A comment rather than an event, so a reader counting
// output does not count it.
func (s *sseWriter) beat() {
	s.mu.Lock()
	defer s.mu.Unlock()
	_, _ = fmt.Fprint(s.w, ": beat\n\n")
	if s.flusher != nil {
		s.flusher.Flush()
	}
}

func (s *sseWriter) stream(name string) io.Writer {
	return streamOf{parent: s, name: name}
}

type streamOf struct {
	parent *sseWriter
	name   string
}

func (o streamOf) Write(p []byte) (int, error) {
	// Reported as written whatever the connection did. The command is running
	// in the cluster and cannot be un-run by a reader that left, and returning
	// an error here would abort it half way — which is the worst of both.
	o.parent.event(o.name, map[string]string{"text": string(p)})
	return len(p), nil
}

func inClusterExec() (Execer, error) {
	cfg, err := rest.InClusterConfig()
	if err != nil {
		return nil, fmt.Errorf("running a command needs the cluster this pod is in: %w", err)
	}
	clients, err := kubernetes.NewForConfig(cfg)
	if err != nil {
		return nil, fmt.Errorf("building the cluster client: %w", err)
	}
	return clusterExec{clients: clients, cfg: cfg}, nil
}

// clusterExec runs commands through the Kubernetes API.
type clusterExec struct {
	clients kubernetes.Interface
	cfg     *rest.Config
}

// NewClusterExec builds an Execer over a Kubernetes client, so something
// outside this package can fill the seam.
func NewClusterExec(clients kubernetes.Interface, cfg *rest.Config) Execer {
	return clusterExec{clients: clients, cfg: cfg}
}

func (c clusterExec) Exec(ctx context.Context, t ExecTarget, stdout, stderr io.Writer) (int, error) {
	pods, err := c.clients.CoreV1().Pods(t.Namespace).List(ctx, metav1.ListOptions{
		LabelSelector: podSelector(t.App),
	})
	if err != nil {
		return 0, fmt.Errorf("listing the pods: %w", err)
	}
	if len(pods.Items) == 0 {
		return 0, ErrNoPods
	}
	pod, err := pickPod(pods.Items)
	if err != nil {
		return 0, err
	}
	container, err := pickContainer(pod, t.Container)
	if err != nil {
		return 0, err
	}

	req := c.clients.CoreV1().RESTClient().Post().
		Resource("pods").Namespace(pod.Namespace).Name(pod.Name).SubResource("exec").
		VersionedParams(&corev1.PodExecOptions{
			Container: container,
			Command:   t.Command,
			Stdout:    true,
			Stderr:    true,
		}, scheme.ParameterCodec)

	exec, err := remotecommand.NewSPDYExecutor(c.cfg, http.MethodPost, req.URL())
	if err != nil {
		return 0, fmt.Errorf("preparing the command: %w", err)
	}
	err = exec.StreamWithContext(ctx, remotecommand.StreamOptions{
		Stdout: stdout, Stderr: stderr,
	})
	// A non-zero exit is the command's answer, not a transport failure. It
	// arrives here as an error and has to stop being one, or every migration
	// that legitimately exits 1 reads as a broken platform.
	var coded utilexec.CodeExitError
	if errors.As(err, &coded) {
		return coded.Code, nil
	}
	if err != nil {
		return 0, err
	}
	return 0, nil
}

// pickPod chooses the pod a command will run in.
//
// Running AND Ready, not merely Running. A pod that has started its container
// but is failing its readiness probe is one the platform is deliberately
// keeping traffic away from, and it is usually mid-crash-loop — running a
// migration in it is running it in the pod most likely to be killed while the
// command is in flight.
//
// When nothing qualifies, the refusal names every pod and what it is doing.
// "No pod is running" on its own is the message that sent somebody to look at
// a Deployment when the answer was an image that would not pull.
func pickPod(pods []corev1.Pod) (corev1.Pod, error) {
	sorted := make([]corev1.Pod, len(pods))
	copy(sorted, pods)
	// By name, so that two identical clusters pick the same pod and a caller
	// running the same command twice does not get two different containers for
	// reasons the API server's list order decides.
	slices.SortFunc(sorted, func(a, b corev1.Pod) int {
		return strings.Compare(a.Name, b.Name)
	})

	for _, pod := range sorted {
		if pod.Status.Phase == corev1.PodRunning && pod.DeletionTimestamp == nil && isReady(pod) {
			return pod, nil
		}
	}

	states := make([]string, 0, len(sorted))
	for _, pod := range sorted {
		states = append(states, pod.Name+" ("+podState(pod)+")")
	}
	return corev1.Pod{}, fmt.Errorf("%w: %s", errNoRunnablePod, strings.Join(states, ", "))
}

func isReady(pod corev1.Pod) bool {
	for _, c := range pod.Status.Conditions {
		if c.Type == corev1.PodReady {
			return c.Status == corev1.ConditionTrue
		}
	}
	return false
}

// podState is the shortest true sentence about why a pod cannot take a command.
//
// The container's waiting reason first, because it is the specific one:
// CrashLoopBackOff, ImagePullBackOff and ContainerCreating all present as a
// pod that is not Running, and they are three different problems with three
// different fixes.
func podState(pod corev1.Pod) string {
	if pod.DeletionTimestamp != nil {
		return "terminating"
	}
	for _, cs := range pod.Status.ContainerStatuses {
		if w := cs.State.Waiting; w != nil && w.Reason != "" {
			return w.Reason
		}
	}
	if pod.Status.Phase == corev1.PodRunning && !isReady(pod) {
		return "running but not ready"
	}
	return string(pod.Status.Phase)
}

// pickContainer chooses which container in the pod, or refuses to.
//
// A pod with one container needs no argument and a pod with several gets no
// guess. The first container is not the app: a workload with a sidecar puts
// the interesting one wherever the template happens to list it, and a
// migration run in a log shipper fails in a way that looks like the migration
// being wrong.
func pickContainer(pod corev1.Pod, want string) (string, error) {
	names := make([]string, 0, len(pod.Spec.Containers))
	for _, c := range pod.Spec.Containers {
		names = append(names, c.Name)
	}
	if want != "" {
		if slices.Contains(names, want) {
			return want, nil
		}
		return "", fmt.Errorf("%w %q in pod %s; it has %s",
			errNoSuchContainer, want, pod.Name, strings.Join(names, ", "))
	}
	switch len(names) {
	case 0:
		// Not reachable through the API server, which rejects a pod with no
		// containers. Answered anyway, because the alternative is indexing
		// names[0] on the day something else builds a Pod value.
		return "", fmt.Errorf("pod %s declares no containers", pod.Name)
	case 1:
		return names[0], nil
	default:
		return "", fmt.Errorf("%w (%s); say which with {\"container\": \"...\"}",
			errAmbiguousContainer, strings.Join(names, ", "))
	}
}

File: panel/assets/exec.js
// Running one command in a container, and reading what it wrote.
//
// The half of exec that a person can reach. The endpoint landed without it, and
// an API-only exec closes none of the gap it was built to close: "get a
// kubeconfig" is the answer it was supposed to replace.
//
// Nothing here runs on load. app.js decides when an app is selected and calls
// mountExec; the two guards at the bottom are the ones every view in this
// directory carries.

// fetch and a reader, and not EventSource — which is the opposite of the
// decision logs.js made, for two reasons that both point this way.
//
// The small one: EventSource only issues GET, and this endpoint is a POST
// because the command is a body rather than a query.
//
// The large one: EventSource reconnects, and reconnecting here would re-run the
// command. logs.js retries five times with backoff because a dropped log stream
// costs nothing to reopen — the lines are still there. A dropped exec stream is
// a command that has already run, possibly a migration, and asking for it again
// is asking for it twice. So this connects once, and when the connection breaks
// it says so instead of trying to be helpful.
const MAX_OUTPUT = 200000;

// The states this view draws, kept apart on purpose.
//
// metrics.js sets the standard this follows: three honest answers about memory
// are three screens, and an empty cell under a heading called "Memory" is a
// claim about memory. The same applies here and the trap is sharper, because
// the failure that looks like nothing is the expensive one.
//
//   idle     nothing has been run
//   running  in flight
//   exited   it ran and returned a code — 0 or not, this is a result
//   refused  it could not be run, and the server said why
//   lost     the connection broke while it was running
//
// "exited" and "refused" are the two that get folded together by every simpler
// version of this file, and folding them is wrong in both directions: a
// migration that exits 1 ran and printed why, and a command refused because
// the pod is crash-looping never touched anything.
//
// "lost" exists because of what it must not look like. The output pane going
// quiet is exactly what a command that printed nothing looks like, and the
// difference is that this one may have changed the database. ingress-nginx
// closes a stream silent for 60 seconds; the server heartbeats inside that, so
// silence here is a real break rather than a slow command.
function execView(state) {
  switch (state.phase) {
    case "idle":
      return { state: "idle", text: "nothing has been run yet", ok: null, ran: false };
    case "running":
      return { state: "running", text: "running…", ok: null, ran: true };
    case "exited": {
      const code = state.code;
      return {
        state: "exited",
        // The server's number, said plainly. A non-zero exit is the command's
        // answer and not a platform failure, so it does not get the word
        // "failed" — it gets its code.
        text: code === 0 ? "the command exited 0" : `the command exited ${code}`,
        ok: code === 0,
        ran: true,
        code,
      };
    }
    case "refused":
      return {
        // The server's own sentence, unedited. It names which of five refusals
        // this was — nothing deployed, no pod that can take a command (and
        // which state each pod is in), more than one container, no such
        // container, no command — and the panel has nothing to add to any of
        // them. Rewriting them here would put the reason behind a second
        // vocabulary that has to be kept in step with the first.
        state: "refused",
        text: state.detail || "the command could not be run",
        ok: false,
        ran: false,
      };
    case "lost":
      return {
        state: "lost",
        text:
          state.detail ||
          "the connection broke while the command was running. It may have finished; " +
          "its output stops here.",
        ok: false,
        // The point of this whole state. Something ran, and what it did is not
        // on this screen.
        ran: true,
      };
    default:
      return { state: state.phase, text: state.phase, ok: null, ran: false };
  }
}

// What a refusal before the stream says.
//
// These arrive as a status and a body rather than as an event, because they
// happen before the first byte: not signed in, not the owner, no such app, an
// install with no cluster to reach. The server's detail is printed whenever
// there is one — catalog.js does the same, and for the same reason.
function refusalText(status, body) {
  const detail = body && typeof body.detail === "string" ? body.detail : "";
  if (detail) return detail;
  switch (status) {
    case 403:
      return "running a command is owner-only, and this account is not an owner";
    case 501:
      return "this installation cannot run commands in the cluster";
    default:
      return `the command was refused with ${status} and no detail`;
  }
}

// Splitting what somebody typed into the argv the endpoint takes.
//
// Quote-aware, because the first thing anyone types is `psql -c "select 1"` and
// splitting that on spaces sends four arguments where two were meant. An
// unbalanced quote is refused by name rather than guessed at: closing it
// silently would run a command the reader did not write.
//
// What this deliberately does not do is wrap the input in `sh -c`. That would
// make every command depend on the image having a shell, which a distroless
// image does not — and the failure would arrive as the kubelet complaining
// about `sh`, naming neither the image nor the assumption.
function parseCommand(text) {
  const argv = [];
  let current = "";
  let quote = null;
  let has = false;
  for (const ch of String(text)) {
    if (quote) {
      if (ch === quote) quote = null;
      else current += ch;
      continue;
    }
    if (ch === '"' || ch === "'") {
      quote = ch;
      has = true;
      continue;
    }
    if (/\s/.test(ch)) {
      if (has || current) argv.push(current);
      current = "";
      has = false;
      continue;
    }
    current += ch;
  }
  if (quote) throw new Error(`the command has an unclosed ${quote} quote`);
  if (has || current) argv.push(current);
  return argv;
}

// SSE framing, by hand, because fetch hands back bytes and not events.
//
// This is the part logs.js declined to write and was right to decline: framing
// a stream by hand is where hand-written stream readers go wrong. It is here
// because EventSource cannot POST, so the choice was this or a GET that carries
// a command in its query string — where it would land in every access log
// between the browser and the pod, which is the one place the server has
// already gone out of its way not to put it.
//
// Returns the complete events and whatever is left over, because a chunk
// boundary falls wherever TCP decides and is nearly never a frame boundary.
// A line beginning with ':' is a comment — the server's heartbeat — and is
// dropped here rather than surfaced, so that a slow command does not look like
// a talkative one.
function parseFrames(buffer) {
  const events = [];
  let rest = String(buffer);
  let at;
  while ((at = rest.indexOf("\n\n")) !== -1) {
    const frame = rest.slice(0, at);
    rest = rest.slice(at + 2);
    let name = "message";
    const data = [];
    for (const line of frame.split("\n")) {
      if (line.startsWith(":")) continue;
      if (line.startsWith("event:")) name = line.slice(6).trim();
      else if (line.startsWith("data:")) data.push(line.slice(5).trim());
    }
    if (!data.length) continue;
    let payload;
    try {
      payload = JSON.parse(data.join("\n"));
    } catch {
      // A frame this file cannot read is reported rather than dropped. Dropping
      // it would take output off the screen and leave the reader believing the
      // command printed less than it did.
      events.push({ event: "stderr", data: { text: data.join("\n") } });
      continue;
    }
    events.push({ event: name, data: payload });
  }
  return { events, rest };
}

// Appending with a ceiling, for the reason logs.js has one: a command is
// entitled to print more than a page can hold.
function appendOutput(text, chunk, max = MAX_OUTPUT) {
  const next = text + chunk;
  return next.length <= max ? next : next.slice(next.length - max);
}

// mountExec renders the view into el and returns a stop function.
//
// stop abandons the reader. It does not and cannot stop the command: it is
// running in a container on the other side of an API server, and the honest
// thing is to say the output was abandoned rather than to imply it was
// cancelled.
function mountExec(el, base, options = {}) {
  const { fetcher = globalThis.fetch, max = MAX_OUTPUT } = options;
  const doc = el.ownerDocument;

  const form = doc.createElement("form");
  form.className = "exec-form";

  const input = doc.createElement("input");
  input.type = "text";
  input.name = "command";
  input.id = "exec-command";
  input.placeholder = "rails db:migrate";
  // No autocomplete and no history, deliberately, and the same rule the server
  // applies to its log: a command line is where people put a password. The
  // server records the program and never the arguments; a browser that
  // remembers this field would undo that by keeping the arguments on disk, on
  // the reader's machine, for the next person at it.
  input.setAttribute("autocomplete", "off");

  const container = doc.createElement("input");
  container.type = "text";
  container.name = "container";
  container.id = "exec-container";
  container.placeholder = "container (only if the pod has several)";
  container.setAttribute("autocomplete", "off");

  const run = doc.createElement("button");
  run.type = "submit";
  run.textContent = "Run";

  const status = doc.createElement("p");
  status.className = "muted";
  const pre = doc.createElement("pre");
  pre.className = "exec-output";
  pre.setAttribute("aria-live", "polite");

  form.replaceChildren(input, container, run);
  el.replaceChildren(form, status, pre);

  let output = "";
  let abort = null;
  let stopped = false;

  const paint = (state) => {
    const view = execView(state);
    status.textContent = view.text;
    status.dataset.ok = String(view.ok);
    status.dataset.state = view.state;
    // Marked on the pane as well, so the stylesheet can make "the output stops
    // here" look unlike a command that simply printed nothing.
    pre.dataset.state = view.state;
  };

  const draw = () => {
    pre.textContent = output;
    pre.scrollTop = pre.scrollHeight;
  };

  paint({ phase: "idle" });

  form.addEventListener("submit", async (event) => {
    event.preventDefault();
    if (stopped || abort) return;

    let argv;
    try {
      argv = parseCommand(input.value);
    } catch (err) {
      paint({ phase: "refused", detail: err.message });
      return;
    }
    if (!argv.length) {
      paint({ phase: "refused", detail: "type a command first" });
      return;
    }

    output = "";
    draw();
    paint({ phase: "running" });
    run.disabled = true;
    abort = new AbortController();

    try {
      const response = await fetcher(execUrl(base), {
        method: "POST",
        headers: { "Content-Type": "application/json" },
        body: JSON.stringify(
          container.value ? { command: argv, container: container.value } : { command: argv },
        ),
        signal: abort.signal,
      });

      if (!response.ok) {
        let body = null;
        try {
          body = await response.json();
        } catch {
          body = null;
        }
        paint({ phase: "refused", detail: refusalText(response.status, body) });
        return;
      }

      const finished = await readStream(response, {
        onText: (text) => {
          output = appendOutput(output, text, max);
          draw();
        },
      });
      paint(finished);
    } catch (err) {
      if (stopped) return;
      // Everything that reaches here is a broken connection rather than a
      // refusal: the request was accepted, so the command is running or has
      // run. This is the state that must not read as silence.
      paint({ phase: "lost" });
    } finally {
      run.disabled = false;
      abort = null;
    }
  });

  return function stop() {
    stopped = true;
    if (abort) abort.abort();
  };
}

// readStream turns the body into the terminal state of one run.
//
// It returns rather than throws for the outcomes the server described, so that
// the caller does not have to tell an exception carrying a server sentence
// apart from one carrying a socket error.
async function readStream(response, { onText }) {
  const reader = response.body.getReader();
  const decode = new TextDecoder();
  let buffer = "";
  let ended = null;

  for (;;) {
    const { value, done } = await reader.read();
    if (done) break;
    buffer += decode.decode(value, { stream: true });
    const { events, rest } = parseFrames(buffer);
    buffer = rest;
    for (const frame of events) {
      switch (frame.event) {
        case "stdout":
        case "stderr":
          onText(frame.data.text || "");
          break;
        case "exit":
          ended = { phase: "exited", code: frame.data.code };
          break;
        case "error":
          ended = { phase: "refused", detail: frame.data.message };
          break;
        default:
          break;
      }
    }
  }
  // The stream finished without saying how. The command was accepted, so
  // something ran and this page does not know what it returned.
  return ended || { phase: "lost" };
}

function execUrl(base) {
  return `${base}/exec`;
}

// The page's entry point. app.js has no imports — the panel has no build step —
// so this is how it reaches this file:
//
//   index.html:  <script src="/exec.js"></script>   before app.js
//   app.js:      window.damgaExec.mountExec(box, base(), {});
if (typeof window !== "undefined") {
  window.damgaExec = { mountExec, execUrl, execView, refusalText, parseCommand, parseFrames, appendOutput };
}

if (typeof module !== "undefined") {
  module.exports = {
    mountExec, execUrl, execView, refusalText, parseCommand, parseFrames, appendOutput, MAX_OUTPUT,
  };
}
